File: 2015/22_WizardSimulator20XX/wizsim.js
/* eslint-disable linebreak-style */
// ======================================================================
// Wizard Simulator 20XX
//   Advent of Code 2015 Day 22 -- Eric Wastl -- https://adventofcode.com
//
// JavaScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           w i z s i m . j s
//
// A solver for the Advent of Code 2015 Day 22 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------
const spells = require('./spells');
const player = require('./player');
const battle = require('./battle');

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------
const WIZARD_TEXT = ['Hit Points: 50', 'Mana: 500'];
const BOSS_TEXT = ['Hit Points: 50', 'Damage: 10'];

// ======================================================================
//                                                                 Wizsim
// ======================================================================

class Wizsim {
  // Object for Wizard Simulator 20XX

  constructor(options) {
    // Create a Wizsim object

    // 1. Set the initial values
    this.text = options.text === undefined ? null : options.text;
    this.part2 = options.part2 === undefined ? false : options.part2;
    this.player = options.player === undefined ? new player.Player({
      part2: this.part2, name: 'Wizard', text: WIZARD_TEXT,
    }) : options.player;
    this.boss = options.boss === undefined ? new player.Player({
      part2: this.part2, name: 'Boss', text: BOSS_TEXT,
    }) : options.boss;
    this.spells = options.spells === undefined ? new spells.Spells({
      part2: this.part2,
    }) : options.spells;
    // 2. Process text (if any)
    if (this.text !== null) {
      this.boss = new player.Player({ text: this.text });
    }
  }

  leastMana(verbose, limit) {
    // 1. Start least mana with more than we have
    let result = 9999999999999999999;
    const original = result;
    const cast = ['Zero', null];
    let rounds = 0;
    // 2. Start a battle
    const conflict = new battle.Battle({
      part2: this.part2,
      player: this.player,
      boss: this.boss,
      spells: this.spells,
    });
    // 3. We need a brain
    function playerBrain(battleArg) {
      let name = battleArg.spells.next(cast[battleArg.round()]);
      let looking = true;
      const active = Object.keys(battleArg.active);
      while (looking) {
        if (name === null) {
          name = 'None';
          looking = false;
        } else {
          const need = battleArg.spells.cost(name);
          if (need === null) {
            name = 'Unknown';
            looking = false;
          } else if (!(name in active)) {
            if (battleArg.player.mana >= need) {
              looking = false;
            }
          }
        }
        if (looking) name = battleArg.spells.next(name);
      }
      return name;
    }
    // 4. Loop until we have exhausted the solutions
    let solving = true;
    while (solving) {
      // 5. Start a round and do the player turn
      let nextSpell = false;
      conflict.newRound();
      // eslint-disable-next-line no-console
      if (verbose) console.log(`Starting round ${conflict.round()} wizard = ${conflict.player.hitPoints}/${conflict.player.mana}  boss = ${conflict.boss.hitPoints}`);
      rounds += 1;
      if (rounds % 10000000 === 0) {
        // eslint-disable-next-line no-console
        console.log(`${rounds} rounds`);
        // eslint-disable-next-line no-console
        console.log(`Spells: ${cast.join(' - ')}`);
        // eslint-disable-next-line no-console
        console.log(`Active = ${conflict.active.join()}`);
      }
      if (limit > 0 && rounds > limit) {
        // eslint-disable-next-line no-console
        if (verbose) console.log(`Reached limit of ${limit} rounds`);
        solving = false;
      } else {
        if (conflict.player.hitPoints < 1) {
          conflict.spell = 'None';
        } else {
          conflict.playerTurn(playerBrain);
        }
        cast[conflict.round()] = conflict.spell;
        // eslint-disable-next-line no-console
        if (verbose) console.log(`Round ${conflict.round()} Player cast ${conflict.spell}`);
        // 5a. If the player is no more, redo this round
        if (conflict.player.hitPoints < 1) {
          // eslint-disable-next-line no-console
          if (verbose) console.log(`Player cast ${conflict.spell} but is no more`);
          nextSpell = true;
          // 5b. if the boss, is no more, check the result and redo this round
        } else if (conflict.boss.hitPoints < 1) {
          // eslint-disable-next-line no-console
          if (verbose) console.log(`Player cast ${conflict.spell} and the boss is no more`);
          if (conflict.player.spent < result) {
            result = conflict.player.spent;
            // eslint-disable-next-line no-console
            console.log(`Rounds: ${rounds}, Setting result to ${result}`);
            // eslint-disable-next-line no-console
            console.log(`Spells: ${cast.join(' - ')}`);
          }
          nextSpell = true;
        } else {
          // 6. Give the boss a turn
          conflict.bossTurn();
          // 6a. If the player is no more, redo this round
          if (conflict.player.hitPoints < 1) {
            // eslint-disable-next-line no-console
            if (verbose) console.log(`Player died after boss's turn in round ${conflict.round()}`);
            nextSpell = true;
            // 6b. if the boss, is no more, check the result and redo this round
          } else if (conflict.boss.hitPoints < 1) {
            // eslint-disable-next-line no-console
            if (verbose) console.log(`boss died after boss's turn in round ${conflict.round()}`);
            if (conflict.player.spent < result) {
              result = conflict.player.spent;
              // eslint-disable-next-line no-console
              console.log(`Rounds: ${rounds}, Setting result to ${result}`);
              // eslint-disable-next-line no-console
              console.log(`Spells: ${cast.join(' - ')}`);
            }
            nextSpell = true;
          }
        }
        // 7. We may have to redo this round, or even back up to the last round
        if (conflict.player.spent > result) {
          // eslint-disable-next-line no-console
          if (verbose) console.log(`Player has spent more than ${result} mana`);
          nextSpell = true;
        }
        if (nextSpell) {
          if (conflict.spell !== 'None' && conflict.spells.next(conflict.spell)) {
            // eslint-disable-next-line no-console
            if (verbose) console.log(`Redoing round ${conflict.round()}`);
            conflict.undo();
          } else {
            conflict.undo();
            cast.pop();
            // eslint-disable-next-line no-console
            if (verbose) console.log(`Backing up to round ${conflict.round()}`);
            conflict.undo();
            // 8. If we have backed up to the start, we are done
            if (conflict.round() === -1) {
              // eslint-disable-next-line no-console
              if (verbose) console.log(`That's all folks at round ${conflict.round()}`);
              solving = false;
            }
          }
        } else {
          cast.push(null);
        }
      }
    }
    // eslint-disable-next-line no-console
    if (verbose) console.log(`Result = ${result}, original = ${original}`);
    // 9. Return the least amount of mana to kill the boss(or null if we never we able to do it)
    if (result < original) {
      return result;
    }
    return null;
  }

  partOne(options) {
    // Returns the solution for part one
    // What is the least amount of mana you can spend and still win the fight?
    // 0. Function arguments
    // eslint-disable-next-line no-unused-vars
    const verbose = options.verbose === undefined ? false : options.verbose;
    // eslint-disable-next-line no-unused-vars
    const limit = options.limit === undefined ? 0 : options.limit;

    // 1. Return the solution for part one
    return this.leastMana(verbose, limit);
  }

  partTwo(options) {
    // Returns the solution for part two

    // 0. Function arguments
    // eslint-disable-next-line no-unused-vars
    const verbose = options.verbose === undefined ? false : options.verbose;
    // eslint-disable-next-line no-unused-vars
    const limit = options.limit === undefined ? 0 : options.limit;

    // 1. Return the solution for part two
    return this.leastMana(verbose, limit);
  }
}

module.exports.Wizsim = Wizsim;
// ======================================================================
// end                        w i z s i m . j s                       end
// ======================================================================

File: 2015/22_WizardSimulator20XX/wizsim.py
# ======================================================================
# Wizard Simulator 20XX
#   Advent of Code 2015 Day 22 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         w i z s i m . p y
# ======================================================================
"A solver for the Advent of Code 2015 Day 22 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
from collections import deque

import spells
import state
import player

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
WIZARD_TEXT = [
    "Hit Points: 50",
    "Mana: 500",
]

BOSS_TEXT = [
    "Hit Points: 13",
    "Armor: 8",
]

MAX_MANA = 2000


# ======================================================================
#                                                                 Wizsim
# ======================================================================


class Wizsim(object):   # pylint: disable=R0902, R0205
    "Object for Wizard Simulator 20XX"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.spells = spells.Spells(text=spells.SPELLS)
        self.states = deque([])

    def initial_state(self, wizard_text=None, boss_text=None):
        "Create the initial game state"

        # 1. Create the two combatants
        if wizard_text is None:
            wizard = player.Player(name="Wizard", text=WIZARD_TEXT, part2=self.part2)
        else:
            wizard = player.Player(name="Wizard", text=wizard_text, part2=self.part2)
        if boss_text is None:
            if self.text is None:
                boss = player.Player(name="Boss", text=BOSS_TEXT, part2=self.part2)
            else:
                boss = player.Player(name="Boss", text=self.text, part2=self.part2)
        else:
            boss = player.Player(name="Boss", text=boss_text, part2=self.part2)

        # 2. Create and save the initial state
        self.states = deque([state.State(wizard=wizard, boss=boss)])

    def least_mana(self):
        "Find the spells that will defeat the boss with the lease expendature of mana"

        # 1. Start assumne that it costs plenty
        result = MAX_MANA
        trace = []

        # 2. Loop while there are states to explore
        while len(self.states) > 0:

            # 3. Get the newest state (depth first)
            utah = self.states.pop()

            # 4. If the boss is dead, check mana usage
            if utah.boss.is_dead():
                print("killed him on turn %d cost %d" % (utah.turn, utah.wizard["used"]))
                if utah.wizard["used"] < result:
                    result = utah.wizard["used"]
                    trace = utah.trace
                continue

            # 5. If the wizard is dead, get a differnt state
            if utah.wizard.is_dead():
                continue

            # 6. Ignore this state if we have already used more than the "best" solution
            if utah.wizard["used"] >= result:
                continue

            # 7. Is there another spell to try? If not, try something else
            spls = utah.another(self.spells)
            if len(spls) == 0:
                continue
            #print("utah", result, utah, spls)

            # 8. Create a new state where we have cast the spell
            ohio = self.turn(utah, self.spells.named(spls[0]))
            if ohio.wizard["used"] >= result:
                continue
            ohio.tried = set()
            self.states.append(ohio)

            # 9. Remember that we used this one and put back for next time
            utah.tried.add(spls[0])
            self.states.appendleft(utah)

        # 10. Return the least amout of mana used
        print(trace)
        return result

    def turn(self, current, spl):
        "Create a new state after the wizard casts the spell and the Boss attacks"

        # 1. Clone the current state and increase the turn number
        ohio = current.clone()
        ohio.turn += 1

        # 2. If "Hard" mode, decrease the player's hit points by one
        if self.part2:
            ohio.wizard.wack()
            if ohio.wizard.is_dead():
                return ohio

        # 3. Give active spells a chance
        ohio.active.effects(ohio.wizard, ohio.boss)

        # 4. If either are dead, that is the end of it
        if ohio.wizard.is_dead() or ohio.boss.is_dead():
            return ohio

        # 4. Cast the spell
        active = spl.cast(ohio.wizard, ohio.boss)
        ohio.cast(spl.name)

        # 5. If a delayed spell, add it to the active set
        if active is not None:
            ohio.active.activate(active)

        # 6. Give active spells a chance
        ohio.active.effects(ohio.wizard, ohio.boss)

        # 7. If either are dead, that is the end of it
        if ohio.wizard.is_dead() or ohio.boss.is_dead():
            return ohio

        # 8. Let the Boss have a go
        ohio.wizard.defend(ohio.boss.attack())

        # 9. Return the new state
        return ohio

    def part_one(self, verbose=False, limit=0):
        "Returns the solution for part one"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part one
        self.initial_state()
        return self.least_mana()

    def part_two(self, verbose=False, limit=0):
        "Returns the solution for part two"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part two
        self.initial_state()
        return self.least_mana()


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                        w i z s i m . p y                       end
# ======================================================================
